lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2557,8 +2557,35 @@ InterfaceTypeRequest::evaluate(Evaluator &eval, ValueDecl *D) const {
       AFD->getParameters()->getParams(argTy);
 
       maybeAddParameterIsolation(infoBuilder, argTy);
-      infoBuilder = infoBuilder.withAsync(AFD->hasAsync());
+
+      auto &ctx = AFD->getASTContext();
+
+      // Check for explicit global actor attribute
+      auto isolationFromAttr = getIsolationFromAttributes(AFD, false);
+
+      // Main.main() and Main.$main are implicitly MainActor-protected.
+      // Any other isolation is an error.
+      if (FuncDecl *fd = dyn_cast<FuncDecl>(AFD)) {
+        std::optional<ActorIsolation> mainIsolation =
+            getActorIsolationForMainFuncDecl(fd);
+        if (mainIsolation) {
+          if (isolationFromAttr && isolationFromAttr->isGlobalActor()) {
+            if (!areTypesEqual(isolationFromAttr->getGlobalActor(),
+                               mainIsolation->getGlobalActor())) {
+              fd->getASTContext().Diags.diagnose(
+                  fd->getLoc(), diag::main_function_must_be_mainActor);
+            }
+            auto isolation = FunctionTypeIsolation::forGlobalActor(
+                mainIsolation->getGlobalActor());
+            infoBuilder = infoBuilder.withIsolation(isolation);
```

Comment:
If we consider that the `mainIsolation` is the correct one, we should use it directly, `isolation` produced from it would be the same as `mainIsolation` itself because it only ever returns `forGlobalActor` isolation with `MainActor` type.

Reply:
mainIsolation is `ActorIsolation` type but `ASTBuilder::withIsolation()` expects `FunctionTypeIsolation` type as it's parameter

Reply:
Oh actually I see now there is a `ASTExtInfo::withGlobalActor(Type globalActor)` function I can pass `mainIsolation->getGlobalActor()` that will create FunctionTypeIsolation for me

Reply:
Right, should just use `mainIsolation->getGlobalActor())` directly

Reply:
updated